src/kbmod/search/raw_image_eigen.cpp

```diff
@@ -0,0 +1,767 @@
+    fits_write_img(fptr, TFLOAT, 1, get_npixels(), image.data(), &status);
+    fits_report_error(stderr, status);
+
+    // Add the basic header data.
+    fits_update_key(fptr, TDOUBLE, "MJD", &obstime, "[d] Generated Image time", &status);
+    fits_report_error(stderr, status);
+
+    fits_close_file(fptr, &status);
+    fits_report_error(stderr, status);
+  }
+
+
+  void RawImageEigen::append_layer_to_file(const std::string& filename) {
+    int status = 0;
+    fitsfile* f;
+
+    // Check that we can open the file.
+    if (fits_open_file(&f, filename.c_str(), READWRITE, &status)) {
+      fits_report_error(stderr, status);
+      throw std::runtime_error("Unable to open FITS file for appending.");
+    }
+
+    // This appends a layer (extension) if the file exists)
+    /* Create the primary array image (32-bit float array) */
+    long dimensions[2];
+    dimensions[0] = width;
+    dimensions[1] = height;
+    fits_create_img(f, FLOAT_IMG, 2 /*naxis*/, dimensions, &status);
+    fits_report_error(stderr, status);
+
+    /* Write the array of floats to the image */
+    fits_write_img(f, TFLOAT, 1, get_npixels(), image.data(), &status);
+    fits_report_error(stderr, status);
+
+    // Save the image time in the header.
+    fits_update_key(f, TDOUBLE, "MJD", &obstime, "[d] Generated Image time", &status);
+    fits_report_error(stderr, status);
+
+    fits_close_file(f, &status);
+    fits_report_error(stderr, status);
+  }
+
+
+  // my implementation here, to test performance for this because no way the
+  // optimization from the original one actually can be worth the messiness
+  const float calc_median(std::vector<float>& values) {
+    std::sort(values.begin(), values.end());
+    const auto len = values.size();
+    return len % 2 == 0 ?
+      (values[len/2-1] + values[len/2]) / 2.0f :
+      values[len/2];
+  }
+
+
+  RawImageEigen create_median_image_eigen2(const std::vector<RawImageEigen>& images) {
+    int num_images = images.size();
+    assert(num_images > 0);
+
+    unsigned height = images[0].get_height();
+    unsigned width = images[0].get_width();
+    RawImageEigen result = RawImageEigen(height, width);
+
+    std::vector<float> pixels;
+    pixels.reserve(num_images);
+    for (unsigned y=0; y<height; ++y){
+      for (unsigned x=0; x<width; ++x){
+        int n_unmasked = 0;
+        for (int i=0; i<num_images; ++i){
+          float pix_val = images[i].get_pixel(x, y);
+          if (pix_val != NO_DATA){
+            pixels.push_back(pix_val);
+          }
+        } // for i
+
+        if (n_unmasked > 0){
+          result.image(x, y) = calc_median(pixels);
+        }
+
+      } // for x
+    } // for y
+
+    return result;
+  }
+
+
+  // default impl
+  RawImageEigen create_median_image_eigen(const std::vector<RawImageEigen>& images) {
+    int num_images = images.size();
+    assert(num_images > 0);
+
+    // if we always use get_pixel why is this neccessary?
+    int width = images[0].get_width();
```

Comment:
This just asserts that all the images are the same size. It is not strictly necessary.

Reply:
What prompted the question was that the the `get_pixel` already protects that by returning no data, so whether or not they were the same size or not, the method would've continued working I think. I guess "are we explicitly considering images of different sizes to be an error or are we supposed to include the edge case in future implementations?" should have been a question.